### PR TITLE
Add auto-preview of image as labels are changed

### PR DESCRIPTION
**Pull request by samwilson**
This adds functionality to be able to live-preview translations as
they're changed in the form. To do this, it updates the API to be
able to first receive (via POST) a new set of translations, which
are then added to the SVG file and a PNG rendered out to the
filesystem. Then, when a 2nd request (GET this time) is received
for that rendered PNG file, it's returned. This is done by creating
a temporary unique identifier for the file (based on the submitted
translations).

The front-end is modified to make this first request, and when it
receives the response all it has to do is update the image's src
attribute and the 2nd request happens.

The process of adding these strings back into SvgFile was a bit
cumbersome, so a new method SvgFile::setTranslations() was
added, which handles adding the new text and tspan node info.

Tests were updated, but the broken controller test remains for
the time being.

Bug: https://phabricator.wikimedia.org/T207203

### CODE REVIEW

**Comment by Niharika29** [comment]
I was testing this out for fun and ran into a bug. It repeatedly fails to generate a preview after filling out some number of labels. Return status 431. I tried with [this file](https://commons.wikimedia.org/wiki/File:Polychaeta_anatomy_en.svg). Steps to reproduce:
1. Load that^ file
2. Start filling in translation labels (I used Hebrew to test with). The previews initially generate alright.
3. After a point, it starts returning error code 431 instead of generating a preview.
I didn't yet figure out what causes the failure. Could easily be something wonky with my setup. I used docker.

**Comment by samwilson** [comment]
> After a point, it starts returning error code 431 instead of generating a preview.
That error is for request header fields being too large. Could be something wonky with my setup. I used docker.

This was a result of my weird draft of getting the modified PNG image in one request. It can't be done (because things get too big): not the way I was trying, nor with a cumbersome base64-encoded image.

The new draft returns to a two-part process: first POST the data and get an image identifier, and then subsequently request the image.

**Comment by MusikAnimal** [comment]
This is really cool! I tried my best to break it, and was unsuccessful :)

I'm a little confused how the temporary files get deleted. It says here they get marked for deletion but I don't see any code that would actually perform this: https://github.com/wikimedia/svgtranslate/pull/42/files#diff-3ed671c8f42cfbd26c9069bfb4cb9acdR103

**Comment by samwilson** [comment]
Good point about that comment; it was a hangover from an earlier idea. Removed now.

The cached files are cleaned up thanks to https://github.com/wikimedia/svgtranslate/blob/452f41267eb1c62b7173a5d6c3bac78ffdacbcc3/src/Service/FileCache.php#L93 — only after the cache lifetime though, which at the moment is 5 hours so we might build up quite a pile of temporary files.

**Comment by MusikAnimal** [comment]
> The cached files are cleaned up thanks to
> 
> [svgtranslate/src/Service/FileCache.php](https://github.com/wikimedia/svgtranslate/blob/452f41267eb1c62b7173a5d6c3bac78ffdacbcc3/src/Service/FileCache.php#L93)
> 
> Line 93 in [452f412](/wikimedia/svgtranslate/commit/452f41267eb1c62b7173a5d6c3bac78ffdacbcc3)
> 
>              unlink($fileName); 
> — only after the cache lifetime though, which at the moment is 5 hours so we might build up quite a pile of temporary files.

Got it. So the files only get deleted if a request is made for the said file? As long as it gets deleted one way or another I think it's fine. I doubt our user base will be so big that will have to worry about disk space for 5 hours worth of cached files.

**Comment by samwilson** [comment]
@MaxSem Did you say that you'd noticed something else amiss with this?

File: src/Controller/ApiController.php [hunk 1]
```diff
@@ -31,45 +34,87 @@ public function __construct(FileCache $cache, Renderer $svgRenderer)
...
+    }
 
-        $translations = \GuzzleHttp\json_decode($json, true);
+    /**
+     * Take POST data, write new translations into the SVG file, render a resultant PNG out to the
+     * filesystem, and return an identifier for that PNG file.
+     *
+     * @Route("/api/translate/{filename}/{lang}", name="api_file_request", methods="POST")
+     */
+    public function requestFileWIthTranslations(string $filename, string $lang, Request $request): Response
```
[the remark this answers is not in the document]
Comment (samwilson) [reply src/Controller/ApiController.php]: Fixed. I also added a phpcs check that will catch this type of thing in future.

File: src/Model/Svg/SvgFile.php [hunk 2]
```diff
@@ -430,6 +430,38 @@ public function getInFileTranslations(): array
         return $this->inFileTranslations;
     }
 
+    /**
+     * Set translations for a single language.
+     * @param string $lang Language code of the translations being provided.
+     * @param string[] $translations Array of tspan-IDs to message texts.
+     * @return string[][]
+     */
+    public function setTranslations(string $lang, array $translations): array
```
Comment (mooeypoo) [inline src/Model/Svg/SvgFile.php]: Hm, this is a bit confusing, but it's also an English preference, so I leave that to proper English speakers -- but... "set translation**s**" sound to me like you can set multiple translations to a single file, and to me, "translations" are per language, so multiple translations = multiple languages.

If this is only me, leave as is, but if this is confusing, then you can change it to either 'setTranslation' (singular) that acceps language + key/value map OR set it to 'setTranslationSets' or something. Translation is usually a grouping of langA->langB words for a single language, isn't it?

File: tests/Model/Svg/SvgFileTest.php [hunk 3]
```diff
@@ -476,6 +476,13 @@ public function testSwitchTranslationSetRoundtrip(): void
         $this->assertEquals(str_replace(' ', '', $origXml), str_replace(' ', '', $this->svg->saveToString()));
     }
 
+    public function testSetTranslations(): void
+    {
+        // The test file does not contain Spanish.
+        $this->svg->setTranslations('es', ['tspan2993' => 'FooBar']);
+        static::assertContains('FooBar', $this->svg->saveToString());
```
Comment (mooeypoo) [inline tests/Model/Svg/SvgFileTest.php]: So I'm a bit confused about this test, I think it's not completely clear. We set a single translation with a language, and when we ask for the file, we see that translation... I am still a little confused with the use of plural-translations but that's an aside. I think what would help here (and make the test better) is if you setTranslation() more than once and then verify what you get. I'm a little confused if we're testing the preview functionality (so you only get the latest set language?) or if we're testing the general functionality (so you get all available translations) but either way, setting it twice will resolve that confusion by merely what the test expects.

For example, if we test that only the latest was added (for only the preview case, not caring about previously set languages)
```
 $this->svg->setTranslations('es', ['tspan2993' => 'FooBar']);
 $this->svg->setTranslations('fr', ['tspan2993' => 'BarFoo']);
// Test only the latest counts
static::assertContains('BarFoo', $this->svg->saveToString());
static::assertNotContains('FooBar', $this->svg->saveToString(BarFoo))
```

If `saveToString` does take into account all languages then you can just make sure both of those check on `assertContains`

**Comment by mooeypoo** [comment]
@samwilson sorry, I took a bit longer to review this, and it was already merged; The majority of my comments are about clarity, which I think we should be doing. Can we do this in a followup?

File: tests/Model/Svg/SvgFileTest.php (hunk 3 above)
Comment (samwilson) [reply tests/Model/Svg/SvgFileTest.php]: Multiple languages can be set, but we don't use that functionality. I've added tests to make it a bit clearer.

**Pull request merged**